### DISCUSSION TITLE
Using Sulu with existing content from a different CMS

### DISCUSSION THREAD

**Discussion by brachycera** [sulu/sulu#6671] [Q&A] [answered]
What would be the best practice in Sulu, when I want to migrate content from an existing CMS into Sulu. As stated in the documentation this should be possible.

> You got existing content you want to integrate through DBAL?

I would like to admin the existing content via Sulu and give my content creators the possibility to copy/edit the existing old articles and create/edit new articles via Sulu. The old content comes from a "News-Website" with about 45.000 articles partly with images.

Thanks for any hints!

**Comment by alexander-schranz** [comment, marked as answer]
Hello @brachycera,

Not 100% sure if I understand your usecase. If we did migrate currently another CMS example Drupal to Sulu. I added an additional [`DBAL Connection`](https://symfony.com/doc/current/doctrine/multiple_entity_managers.html) to the other CMS database (mostly it was a SQL dump from prod system). And used the a Symfony Command:

https://symfony.com/doc/current/console.html

To read all articles and create via the document manager the Sulu Articles.

A short snippet can be found here:

Data should be loaded over a \Generator method: https://gist.github.com/alexander-schranz/e526e8cce7755220045132034bbe5687#file-documentmanagerimportcommand-php-L16

And then the document is created or updated: https://gist.github.com/alexander-schranz/e526e8cce7755220045132034bbe5687#file-documentmanagerimportcommand-php-L23 ("createOrUpdate" is nice so you can easily run the command multiple times, I used a mapping table (externalId) to (uuid) to know if I need update (load for update) or create the document).

You can then map your old data to your new sulu template via the getStructure()->bind and setExtension methods. See for this: http://docs.sulu.io/en/latest/reference/components/document-manager/fixtures.html 

This way I imported a lot of data from different Drupals into Sulu. As the drupals I worked with had no structured content just a lot of big text editor content, I did also parse text editors HTML and split it into functional sulu blocks via the [DOMCrawler](https://symfony.com/doc/current/components/dom_crawler.html) of symfony.

**Comment by brachycera** [comment]
Hi @alexander-schranz 
Thanks again for you answer and help, very much appreciated! :thumbsup: :pray: I'm still trying to figure out Sulu and Symfony as I move forward, I only worked with Symfony indirectly via Laravel. 

When I understood you correctly, you create a console command to query all articles from the old (drupal) CMS and use the document manager to presist/save all this articles into Sulu?

Example: https://gist.github.com/brachycera/8f7ce72e4cd70c4d58fc22a839d97e0c

I think one of the puzzle pieces where I need to wrap my head around is, how a single `$dataList` element from your `load()` method should be structured. And I need to learn more about how the `entity manager` and `document manager` are working. :bowtie: FYI the "old" CMS which I'm trying to import to Sulu is a [textpattern](http://www.textpattern.com) installation.

**Reply by alexander-schranz** [reply]
Yes it is kind of that. The data returned by `load` was a \Generator (because of performance) which returns an array. The array can be structured whatever it will be at the end you need to map your data to the sulu data so you will have mostly a method doing that in your case. I mostly tried to already be as near as possible with the sulu format already loading data from the old system. But as said we did also parse HTML converted HTML into several Blocks. At the end ist you just need to know where the data is stored in your old CMS and in which field or fields you want to map it into Sulu. At the begin try to get familiar just how the document manager works and how you can save different field types. Best way to findout how a field type need to be constructed is opening the Template in your Sulu Admin and save it their and in the Browser Network Tab you will see how the data need to be structured for the specific field type. As some could be a little bit tricky as media selection and teasers.

**Reply by brachycera** [reply]
I think the fog on how and were to get started is slowly clearing. Will starting to learn more about the `document manager`. Hopefully i don't have to bother you again, thanks!  :wink:

**Reply by alexander-schranz** [reply]
Always feel free to ask any questions. Will try todo my best to help you.

**Reply by brachycera** [reply]
Hey @alexander-schranz I'm back with a new Question.
I managed to query the articles from the old CMS:

- added a new `external` connection to `doctrine.yaml`
  - added the "database table filter" option to the new connection `schema_filter: ~^textpattern$~` 
- used `bin/adminconsole doctrine:mapping:import App\\Entity annotation --em=external --path=src/Entity` to create a "textpattern" table Entity class
- query articles with `$this->managerRegistry->getManager('external')->getRepository(Textpattern::class)->findAll()`

Yay, so far so good. Next I tried to create a dummy article to learn how to interact with the `DocumentManager`.
This sadly didn't work. I'm not sure if the error is on my side, if it is a bug in sulu or the docs are missing something? I read in the #5970 issue that there was something familiar happening.

To create an example page/article I used the snippet from the `load()` method as seen in the Sulu documentation - https://docs.sulu.io/en/latest/reference/components/document-manager/fixtures.html?highlight=persist#getting-started

What happens is the article/page Node gets created, in the DB table `phpcr_nodes`, along the route Node for the page. BUT the page doesn't show up in the Sulu backend.

What I figured out is that the corresponding Nodes (page+route) with the `workspace_name` value `default` aren't created. When I copy the 2 newly created Nodes directly in the database, and change the `workspace_name` value to `default`. The newly created page appears in the Sulu backend.

On the other hand, when I create an example page without the `$this->documentManager->publish($document, 'en');` method call - only the page Node is created and no corresponding route Node. Also this page doesn't show up in the Sulu Backend.

I hope the description of the problem makes sense :flushed: and you might can help me?

**Reply by brachycera** [reply]
Looks like I found the error by myself. This [stackoverflow answear](https://stackoverflow.com/a/42910547/1423758) brought me on the right track. I tried the to create a document in a "website" controller - to get better debug/stacktrace infos - but the controller runs on the website kernel :confused:. When I ran the code in my Command::class, the new document was created as expected :stuck_out_tongue_winking_eye:. I'm curious where in the sulu codebase did this happen???

**Reply by alexander-schranz** [reply]
@brachycera I personally would not register a `entity_manager` or `orm` for that connection. I would use directly just the dbal connection to query that other database. I think creating a TextPattern entity does in this case not add any advantages specially if you need to import data from different tables is simple SQL (over dbal querybuilder) and array result easier to work with instead of need to create ORM Entity and it for everything.


> I'm curious where in the sulu codebase did this happen???

Sulu basically has a 2 Kernel Setup. While in 1.6 it was more visible in 2.0 it is mostly hidden and in the `Admin` Kernel the default PHPCR connection is to `draft/default` workspace and in the `Website` Kernel the default PHPCR connection is to the `live` workspace. And there is so the limitation that the `Website` can currently not create documents as it does not know anything about the workspace saving where the drafts/admin documents are saved in. This issue will be targetting with the new Content Storage in 3.0 which will be provided by the [SuluContentBundle](https://github.com/sulu/SuluContentBundle).
